src/box/lua/config/applier/box_cfg.lua

```diff
@@ -426,7 +441,82 @@ local function apply(config)
     box.cfg(box_cfg)
 end
 
+local function apply_names_schema_on_replace(old, new)
+    if old == nil and new[1] == 'replicaset_name' then
+        box.broadcast('apply_name', new[2])
```

Comment:
Let's use some less obvious name. For example, `box.apply_name`, or `box.config.*`
It's internal, it shouldn't share format with normal user event names.

Reply:
It's forbidden to create broadcasts, which start from `box.*`, from Lua. For now I've given it a name `box_apply_name`.

If we really want to make it start with `box`, we can expose box_broadcast to `box.internal`. However, I think it's an overkill. All triggers and watcher itself are deleted as soon as all names are set.

I don't really care, so if you insist, we'll expose and use `box.apply_name`

Reply:
I found a major bug in the previous implementation: only one name and only one alert disappeared after setting/getting names. After some investigation I figured out, that broadcasts replaces a value, if it have not been delivered yet:

```lua
tarantool> box.watch('foo', function(_, v) print(v) end)
---
- box.watcher
...

nil
tarantool> box.broadcast('foo', 1); box.broadcast('foo', 2)
---
...

2
```

This is following code:

<details><summary>Code reference</summary>

https://github.com/tarantool/tarantool/blob/952d15827b9a299e6a13a8f0e3a0f3b0c36118cc/src/box/watcher.c#L396-L409

</details>

Firstly, I tried to make everything work inside `on_replace` trigger. The problem out there, is that we cannot call box.cfg there, as the name is not considered set in `on_replace` or `on_commit` trigger. So, setting `repicaset_name` on replica causes `read_only` error, `instance_name` - resubscribe. 

So, I done everything using `box.broadcast`. As I don't have a great desire to rewrite the current implementation of watchers and make them to process every broadcast, and not only the last one, I've done it in lua using `watch_once`. Hope, it's ok.

After that I encountered the problem, that resubscribe still happens, when replica tries to set its instance name. Moreover, replica hangs on that box.cfg during xlog recovering after resubscribe with the following log:

<details><summary>Logs of the replica</summary>

```
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main/110/applier/unix/:./instance-001.iproto I> assigned name instance-002 to replica 2c3e1bcc-6da4-4bfd-8268-e2966ac2ea72
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main/110/applier/unix/:./instance-001.iproto I> assigned name instance-001 to replica 755fa8b0-b92b-4b0d-8713-9e3e4b86cd58
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main I> leaving orphan mode
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main/tarantool.config log.lua:146 W> BROADCASTING INSTANCE: ["instance-002"]
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main/tarantool.config log.lua:146 W> BROADCASTING REPLCIASET: ["instance-002","replicaset-001"]
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main/tarantool.config log.lua:146 W> BROADCASTING INSTANCE: ["instance-002","replicaset-001","instance-001"]
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main I> replicaset name: replicaset-001
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main/128/box.watcher/tarantool.config log.lua:146 W> I GOT: ["instance-002","replicaset-001","instance-001"]
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main/128/box.watcher/tarantool.config log.lua:146 W> NAME: instance-002
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main/128/box.watcher/tarantool.config log.lua:130 W> SETTING BOX.CFG.INSTANCE_NAME
[001] instance-002 | 2023-10-15 23:48:11.028 [215132] main/128/box.watcher I> connecting to 2 replicas
[001] instance-002 | 2023-10-15 23:48:11.029 [215132] main/130/applier/unix/:./instance-002.iproto I> remote master 2c3e1bcc-6da4-4bfd-8268-e2966ac2ea72 at unix/:./instance-002.iproto running Tarantool 3.0.0
[001] instance-002 | 2023-10-15 23:48:11.029 [215132] main/129/applier/unix/:./instance-001.iproto I> remote master 755fa8b0-b92b-4b0d-8713-9e3e4b86cd58 at unix/:./instance-001.iproto running Tarantool 3.0.0
[001] instance-002 | 2023-10-15 23:48:11.030 [215132] main/128/box.watcher I> connected to 2 replicas
[001] instance-002 | 2023-10-15 23:48:11.030 [215132] main/129/applier/unix/:./instance-001.iproto applier.cc:2536 W> need_name: 0
[001] instance-002 | 2023-10-15 23:48:11.030 [215132] main/129/applier/unix/:./instance-001.iproto I> can't join/subscribe
[001] instance-002 | 2023-10-15 23:48:11.030 [215132] main/129/applier/unix/:./instance-001.iproto box.cc:4583 E> ER_CFG: Incorrect value for option 'replication': duplicate connection with the same replica UUID
[001] instance-002 | 2023-10-15 23:48:11.030 [215132] main/129/applier/unix/:./instance-001.iproto I> will retry every 1.00 second
[001] instance-001 | 2023-10-15 23:48:11.030 [215128] relay/unix/:(socket)/101/main coio.c:348 E> SocketError: unexpected EOF when reading from socket, called on fd 39, aka unix/:(socket), peer of unix/:(socket): Broken pipe
[001] instance-001 | 2023-10-15 23:48:11.030 [215128] relay/unix/:(socket)/101/main I> exiting the relay loop
[001] instance-002 | 2023-10-15 23:48:12.031 [215132] main/129/applier/unix/:./instance-001.iproto applier.cc:2536 W> need_name: 0
[001] instance-002 | 2023-10-15 23:48:12.031 [215132] main/129/applier/unix/:./instance-001.iproto I> subscribed
[001] instance-002 | 2023-10-15 23:48:12.031 [215132] main/129/applier/unix/:./instance-001.iproto I> remote vclock {1: 2, 2: 3} local vclock {1: 2, 2: 3}
[001] instance-001 | 2023-10-15 23:48:12.031 [215128] main/127/main I> subscribed replica 2c3e1bcc-6da4-4bfd-8268-e2966ac2ea72 at fd 40, aka unix/:(socket), peer of unix/:(socket)
[001] instance-001 | 2023-10-15 23:48:12.031 [215128] main/127/main I> remote vclock {1: 2, 2: 3} local vclock {1: 2, 2: 3}
[001] instance-002 | 2023-10-15 23:48:12.032 [215132] main/129/applier/unix/:./instance-001.iproto I> RAFT: message {term: 1, state: follower} from 2
[001] instance-001 | 2023-10-15 23:48:12.032 [215128] relay/unix/:(socket)/101/main I> recover from `/tmp/t/001_config-luatest/rs-7vE6L55l2Kep/instance-001-bcqinc2hWUUj/00000000000000000002.xlog'
```

</details>

I still investigate this hang, as I don't understand its reason. If you have any ideas, share them please. Anyway, we should not resubscribe, if name is already set, so I changed the behavior of `box_set_instance_name` a little bit.

At the end of a day, I'm asking you to take a look at commit `config: automatically set missing names` one more time.  I also extended tests, now they work without crutches.

Reply:
> So, I done everything using box.broadcast. As I don't have a great desire to rewrite the current implementation of watchers and make them to process every broadcast, and not only the last one, I've done it in lua using watch_once. Hope, it's ok.

Looks good to me.

This is because without checking that real instance name is already set, `box_set_instance_name` did `box_register_on_master`, which tries to wait for applier to pass "REGISTERED" state, which never happens, since applier knows it doesn't have to register.
https://github.com/tarantool/tarantool/blob/952d15827b9a299e6a13a8f0e3a0f3b0c36118cc/src/box/box.cc#L2230-L2231
That said, you changed `box_set_instance_name` behavior correctly, and I suggest adding a test that there is no hang when setting the already-set name.

Reply:
> It's forbidden to create broadcasts, which start from `box.*`, from Lua. For now I've given it a name `box_apply_name`.
> 
> If we really want to make it start with `box`, we can expose box_broadcast to `box.internal`. However, I think it's an overkill. All triggers and watcher itself are deleted as soon as all names are set.
> 
> I don't really care, so if you insist, we'll expose and use `box.apply_name`

I'd suggest to make the event name start with `box.` so that the event never clashes with any user-created event. Besides, it's our internal event, even though created from lua. It should use the same naming scheme.

You don't have to expose `box_broadcast`, just a slight change to `lbox_broadcast` is needed. Create `lbox_internal_broadcast` and use the common body for it and `lbox_broadcast`. But don't check for names matching `box.*` in `lbox_internal_broadcast`.

Reply:
> I'd suggest to make the event name start with box.

Sure, let's do that

Reply:
> Thank you a lot! It indeed is the problem. Introduced a test, made it a separate commit

Nice, thanks!